### PR TITLE
feature/48/display-building-information

### PR DESCRIPTION
**Pull request by sandraroz**
As a user, I’d like to know additional information for each building so I can quickly access all the basic information for a facility #28 

I'll need help with the UI and adding services to the building since I'm probably going to take over testing at this point. I'm working on adding another unit test with mocking to up the coverage and will move on to automated testing.

### CODE REVIEW

File: app/src/main/java/com/conupods/OutdoorMaps/BuildingDataMap.java [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+                );
+                mData.put(latLng, buildingObj);
+            }
+        } catch (FileNotFoundException e) {
+            throw new FileNotFoundException(e.getMessage());
+        } catch (IOException e) {
+            throw new IOException(e.getMessage());
+        } catch (ParseException e) {
+            throw new Exception(e.getMessage());
+        }
```
Comment (minazenhub) [inline app/src/main/java/com/conupods/OutdoorMaps/BuildingDataMap.java]: you are re-throwing all those exceptions but creating new ones instead. Why? Also you are losing info by turning the parseException into a regular Exception

File: app/src/main/java/com/conupods/OutdoorMaps/BuildingDataMap.java [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+/**
+ * A singleton class that holds hardcoded building data of concordia buildings.
+ * The class parses the data (only once) that is saved in the file ../assets/buildings.json
+ * who was imported from the concordia API manually.
+ */
+public class BuildingDataMap {
+    private static BuildingDataMap mInstance;
+    private HashMap<LatLng, Building> mData;
+
+    private BuildingDataMap() throws Exception {
```
Comment (minazenhub) [inline app/src/main/java/com/conupods/OutdoorMaps/BuildingDataMap.java]: Create a specific Exception, Exception is the worst most generic type

File: app/src/main/java/com/conupods/OutdoorMaps/BuildingDataMap.java [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+        if (mInstance == null) {
+            try {
+                mInstance = new BuildingDataMap();
+            } catch (Exception e) {
+                e.printStackTrace();
+                return null;
+            }
+        }
+        return mInstance;
+    }
```
Comment (minazenhub) [inline app/src/main/java/com/conupods/OutdoorMaps/BuildingDataMap.java]: Nice job on the singleton!

File: app/src/main/java/com/conupods/OutdoorMaps/BuildingInfoWindow.java [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+                setTextViewContentsNotConcordia(v, marker);
+            }
+        }
+
+        // Returning the updated view
+        return v;
+    }
+
+    private void setTextViewContentsNotConcordia(View v, Marker marker) {
+        TextView marker_title = v.findViewById(R.id.title);
```
Comment (minazenhub) [inline app/src/main/java/com/conupods/OutdoorMaps/BuildingInfoWindow.java]: for objects we use camelCase

File: app/src/main/java/com/conupods/OutdoorMaps/BuildingInfoWindow.java [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+        marker_title.setText(marker.getTitle());
+    }
+
+    private void setTextViewContentsConcordia(View v, Building building) {
+        // Getting reference to the TextView
+        TextView building_campus = v.findViewById(R.id.building_campus);
+        TextView building_longName = v.findViewById(R.id.building_longName);
+        TextView building_code = v.findViewById(R.id.title);
+        TextView building_address = v.findViewById(R.id.building_address);
+
```
Comment (minazenhub) [inline app/src/main/java/com/conupods/OutdoorMaps/BuildingInfoWindow.java]: camelCase

File: app/src/main/java/com/conupods/OutdoorMaps/View/MapsActivity.java [hunk 6]
```diff
@@ -17,26 +15,28 @@
...
 import com.google.android.gms.maps.OnMapReadyCallback;
+import com.google.android.gms.maps.GoogleMap.OnMarkerClickListener;
 import com.google.android.gms.maps.SupportMapFragment;
+import com.google.android.gms.maps.model.Marker;
 import com.google.android.gms.tasks.OnFailureListener;
 import com.google.android.gms.tasks.OnSuccessListener;
 import com.google.android.gms.tasks.Task;
 
-public class MapsActivity extends FragmentActivity implements OnMapReadyCallback {
+public class MapsActivity extends FragmentActivity implements OnMapReadyCallback, OnMarkerClickListener {
```
Comment (minazenhub) [inline app/src/main/java/com/conupods/OutdoorMaps/View/MapsActivity.java]: OnMarkerListener should be in its own class

File: app/src/main/java/com/conupods/OutdoorMaps/BuildingDataMap.java [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+    private static final String TAG = "BUILDING_DATA_MAP";
+
+    private BuildingDataMap() throws BuildingException {
+        // todo possibly change the index to be the marker object and not latlng
+        mData = new HashMap<>();
+        parseBuildingData();
+    }
+
+    public static synchronized BuildingDataMap getInstance() {
+        if (mInstance == null) {
```
Comment (MinaSelim) [inline app/src/main/java/com/conupods/OutdoorMaps/BuildingDataMap.java]: what if you use eager initialization? the synchronized will add some overhead, and we know for a fact that we will need this stuff at the startup of the app. so while the loading screen is running, this code can be ran.

File: app/src/main/java/com/conupods/OutdoorMaps/MarkerClickListener.java [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+        this.mMap = map;
+        this.mBuildingInfoWindow = buildingInfoWindow;
+    }
+
+    @Override
+    public boolean onMarkerClick(Marker marker) {
+        mMap.setInfoWindowAdapter(mBuildingInfoWindow);
+        return false;
+    }
+}
```
Comment (MinaSelim) [inline app/src/main/java/com/conupods/OutdoorMaps/MarkerClickListener.java]: I think this logic should be put inside the mapinitializer class, turning the onMarkerClick into a lambda function. You will have to modify the constructor of the MapInitializer to take in your buildingInformation object.

File: app/src/main/java/com/conupods/OutdoorMaps/BuildingDataMap.java [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+
+/**
+ * A singleton class that holds hardcoded building data of concordia buildings.
+ * The class parses the data (only once) that is saved in the file ../assets/buildings.json
+ * who was imported from the concordia API manually.
+ */
+public class BuildingDataMap {
+    private static BuildingDataMap mInstance;
+    private HashMap<LatLng, Building> mData;
+    private static final String TAG = "BUILDING_DATA_MAP";
```
Comment (domludera) [inline app/src/main/java/com/conupods/OutdoorMaps/BuildingDataMap.java]: shouldn't this follow member variable conventions?

**Comment by minazenhub** [comment]
The code looks great, I cant approve since I didn't go on the branch to manually test it. I think the @conso1e is supposed to be the one to pull it and approve it

File: app/src/main/java/com/conupods/OutdoorMaps/BuildingDataMap.java (hunk 3 above)
Comment (conso1e) [reply app/src/main/java/com/conupods/OutdoorMaps/BuildingDataMap.java]: Can we move this class to the services package? Mike and I need exactly this for the search feature.

**Comment by conso1e** [comment]
Are all of the markers supposed to be added to the map? The display becomes really cluttered. Do we even need them in the first place?

**Comment by conso1e** [comment]
We might want to consider some kind of word wrapping for the text boxes. They become pretty clunky for buildings with long names.
image

**Pull request merged**